third/nmh/man/mhn.man
.\"
.\" %nmhwarning%
.\" $Id: mhn.man,v 1.1.1.1 1999-02-07 18:14:21 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH MHN %manext1% MH.6.8 [%nmhversion%]
.SH NAME
mhn \- display/list/store/cache MIME messages
.SH SYNOPSIS
.in +.5i
.ti -.5i
mhn \%[+folder] \%[msgs] \%[\-file file]
.br
\%[\-part number]... \%[\-type content]...
.br
\%[\-show] \%[\-noshow]
\%[\-list] \%[-nolist]
.br
\%[\-store] \%[\-nostore]
\%[\-cache] \%[\-nocache]
.br
\%[\-headers] \%[\-noheaders]
\%[\-realsize] \%[\-norealsize]
.br
\%[\-serialonly] \%[\-noserialonly]
\%[\-form formfile]
.br
\%[\-pause] \%[\-nopause]
\%[\-auto] \%[\-noauto]
.br
\%[\-rcache policy] \%[\-wcache policy]
\%[\-check] \%[\-nocheck]
.br
\%[\-verbose] \%[\-noverbose]
\%[\-version]
\%[\-help]

.ti .5i
mhn \-build\ file
.br
\%[\-ebcdicsafe] \%[\-noebcdicsafe]
.br
\%[\-rfc934mode] \%[\-norfc934mode]
.in -.5i

.SH DESCRIPTION
MHN SHOULD BE CONSIDERED DEPRECATED.  IT IS RETAINED FOR THE PURPOSE
OF BACKWARD COMPATIBILITY, BUT EVERYONE SHOULD MIGRATE TO USING THE
COMMANDS MHSHOW, MHSTORE, AND MHLIST.  CHECK THE INDIVIDUAL MAN PAGES
FOR DETAILS.

The \fImhn\fR command allows you to display, list, store, or cache the
contents of a MIME (multi-media) messages.

\fImhn\fR manipulates multi-media messages as specified in RFC\-2045
thru RFC\-2049.  Currently \fImhn\fR only supports encodings in message
bodies, and does not support the encoding of message headers as specified
in RFC\-2047.

The switches `\-list', `\-show', `\-store', and `-cache' direct
the operation of \fImhn\fR.  Only one of these switches may be used
at a time.  These switches are used to operate on the content of
each of the named messages.  By using the `\-part' and `\-type'
switches, you may limit the scope of the given operation to particular
subparts (of a multipart content) and/or particular content types.

The switch `\-build' is used to construct a MIME message.  It is
for backward compatibility and instructs \fImhn\fR to execute the
\fImhbuild\fR command.  It is preferred that you use the \fImhbuild\fR
command directly.  See the \fImhbuild\fR(1) man page for details.

The option `\-file\ file' directs \fImhn\fR to use the specified file as
the source message, rather than a message from a folder.  If you specify
this file as \*(lq-\*(rq, then \fImhn\fR will accept the source message
on the standard input.  Note that the file, or input from standard input
should be a validly formatted message, just like any other \fInmh\fR
message.  It should \fBNOT\fR be in mail drop format (to convert a file in
mail drop format to a folder of \fInmh\fR messages, see \fIinc\fR\0(1)).

A part specification consists of a series of numbers separated by dots.
For example, in a multipart content containing three parts, these
would be named as 1, 2, and 3, respectively.  If part 2 was also a
multipart content containing two parts, these would be named as 2.1 and
2.2, respectively.  Note that the `\-part' switch is effective for only
messages containing a multipart content.  If a message has some other
kind of content, or if the part is itself another multipart content, the
`\-part' switch will not prevent the content from being acted upon.

A content specification consists of a content type and a subtype.
The initial list of \*(lqstandard\*(rq content types and subtypes can
be found in RFC\-2046.
.ne 18
A list of commonly used contents is briefly reproduced here:
.sp
.nf
.in +.5i
.ta \w'application  'u
Type	Subtypes
----	--------
text	plain, enriched
multipart	mixed, alternative, digest, parallel
message	rfc822, partial, external-body
application	octet-stream, postscript
image	jpeg, gif, png
audio	basic
video	mpeg
.re
.in -.5i
.fi
.sp
A legal MIME message must contain a subtype specification.
.PP
To specify a content, regardless of its subtype, just use the
name of the content, e.g., \*(lqaudio\*(rq.  To specify a specific
subtype, separate the two with a slash, e.g., \*(lqaudio/basic\*(rq.
Note that regardless of the values given to the `\-type' switch, a
multipart content (of any subtype listed above) is always acted upon.
Further note that if the `\-type' switch is used, and it is desirable to
act on a message/external-body content, then the `\-type' switch must
be used twice: once for message/external-body and once for the content
externally referenced.

.Uh "Checking the Contents"
The `\-check' switch tells \fImhn\fR to check each content for an
integrity checksum.  If a content has such a checksum (specified as a
Content-MD5 header field), then \fImhn\fR will attempt to verify the
integrity of the content.

.Uh "Listing the Contents"
The `\-list' switch tells \fImhn\fR to list the table of contents
associated with the named messages.

The `\-headers' switch indicates that
a one-line banner should be displayed above the listing.  The `\-realsize'
switch tells \fImhn\fR to evaluate the \*(lqnative\*(rq (decoded) format
of each content prior to listing.  This provides an accurate count at
the expense of a small delay.  If the `\-verbose' switch is present, then
the listing will show any \*(lqextra\*(rq information that is present in
the message, such as comments in the Content-Type header.

.Uh "Showing the Contents"
The `\-show' switch tells \fImhn\fR to display the contents of the named
messages.

The headers of each message are displayed with the \fImhlproc\fR
(usually \fImhl\fR), using the standard format file \fImhl.headers\fR.
You may specify an alternate format file with the `\-form formfile'
switch.  If the format file \fImhl.null\fR is specified, then the display
of the message headers is suppressed.

The method used to display the different contents in the messages bodies
will be determined by a \*(lqdisplay string\*(rq.  To find the display
string, \fImhn\fR will first search your profile for an entry of the form:
.sp
.in +.5i
mhn-show-<type>/<subtype>
.in -.5i
.sp
to determine the display string.  If this isn't found, \fImhn\fR
will search for an entry of the form:
.sp
.in +.5i
mhn-show-<type>
.in -.5i
.sp
to determine the display string.

If a display string is found, any escapes (given below) will be expanded.
The result will be executed under \fB/bin/sh\fR, with the standard input
set to the content.
.ne 16
The display string may contain the following escapes:
.sp
.nf
.in +.5i
.ta \w'%F  'u
%a	Insert parameters from Content-Type field
%e	exclusive execution
%f	Insert filename containing content
%F	%e, %f, and stdin is terminal not content
%l	display listing prior to displaying content
%p	%l, and ask for confirmation
%s	Insert content subtype
%d	Insert content description
%%	Insert the character %
.re
.in -.5i
.fi
.sp
.ne 10
For those display strings containing the e- or F-escape, \fImhn\fR will
execute at most one of these at any given time.  Although the F-escape
expands to be the filename containing the content, the e-escape has no
expansion as far as the shell is concerned.

When the p-escape prompts for confirmation, typing INTR (usually
control-C) will tell \fImhn\fR not to display that content.  The p-escape
can be disabled by specifying the switch `\-nopause'.  Further, when
\fImhn\fR is display a content, typing QUIT (usually control-\\) will
tell \fImhn\fR to wrap things up immediately.

Note that if the content being displayed is multipart, but not one of
the subtypes listed above, then the f- and F-escapes expand to multiple
filenames, one for each subordinate content.  Further, stdin is not
redirected from the terminal to the content.

If a display string is not found, \fImhn\fR has several default values:
.sp
.nf
.in +.5i
mhn-show-text/plain: %pmoreproc '%F'
mhn-show-message/rfc822: %pshow -file '%F'
.in -.5i
.fi
.sp
If a subtype of type text doesn't have a profile entry, it will be
treated as text/plain.

\fImhn\fR has default methods for handling multipart messages of subtype
mixed, alternative, parallel, and digest.  Any unknown subtype of type
multipart (without a profile entry), will be treated as multipart/mixed.

If none of these apply, then \fImhn\fR will check to see if the message
has an application/octet-stream content with parameter \*(lqtype=tar\*(rq.
If so, \fImhn\fR will use an appropriate command.  If not, \fImhn\fR
will complain.

.ne 10
Example entries might be:
.sp
.nf
.in +.5i
mhn-show-audio/basic: raw2audio 2>/dev/null | play
mhn-show-image: xv '%f'
mhn-show-application/PostScript: lpr -Pps
.in -.5i
.fi
.sp
Note that when using the f- or F-escape, it's a good idea to use
single-quotes around the escape.  This prevents misinterpretation by
the shell of any funny characters that might be present in the filename.

Finally, \fImhn\fR will process each message serially\0--\0it won't start
showing the next message until all the commands executed to display the
current message have terminated.  In the case of a multipart content
(of any subtype listed above), the content contains advice indicating if
the parts should be displayed serially or in parallel.  Because this may
cause confusion, particularly on uni-window displays, the `\-serialonly'
switch can be given to tell \fImhn\fR to never display parts in parallel.

.Uh "Showing Alternate Character Sets"
Because a content of type text might be in a non-ASCII character
set, when \fImhn\fR encounters a \*(lqcharset\*(rq parameter for
this content, it checks if your terminal can display this character
set natively.  \fIMhn\fR checks this by examining the the environment
variable MM_CHARSET.  If the value of this environment variable is equal
to the value of the charset parameter, then \fImhn\fR assumes it can
display this content without any additional setup.  If this environment
variable is not set, \fImhn\fR will assume a value of \*(lqUS-ASCII\*(rq.
If the character set cannot be displayed natively, then \fImhn\fR will
look for an entry of the form:
.sp
.in +.5i
mhn-charset-<charset>
.in -.5i
.sp
which should contain a command creating an environment to render
the character set.  This command string should containing a single
\*(lq%s\*(rq, which will be filled-in with the command to display the
content.

Example entries might be:
.sp
.in +.5i
mhn-charset-iso-8859-1: xterm -fn '-*-*-medium-r-normal-*-*-120-*-*-c-*-iso8859-*' -e %s
.in -.5i
or
.in +.5i
mhn-charset-iso-8859-1: '%s'
.in -.5i
.sp
The first example tells \fImhn\fR to start \fIxterm\fR and load the
appropriate character set for that message content.  The second example
tells \fImhn\fR that your pager (or other program handling that content
type) can handle that character set, and that no special processing is
needed beforehand.
.sp
Note that many pagers strip off the high-order bit or have problems
displaying text with the high-order bit set.  However, the pager
\fIless\fR has support for single-octet character sets.  The source
to \fIless\fR is available on many ftp sites carrying free software.
In order to view messages sent in the ISO-8859-1 character set using
\fIless\fR,
.ne 9
put these lines in your \&.login file:
.sp
.nf
.in +.5i
setenv LESSCHARSET latin1
setenv LESS "-f"
.in -.5i
.fi
.sp
The first line tells \fIless\fR to use the ISO-8859-1 definition for
determining whether a character is \*(lqnormal\*(rq, \*(lqcontrol\*(lq,
or \*(lqbinary\*(rq.  The second line tells \fIless\fR not to warn you
if it encounters a file that has non-ASCII characters.  Then, simply
set the \fBmoreproc\fR profile entry to \fIless\fR, and it will get
called automatically.  (To handle other single-octet character sets,
look at the \fIless\fR\0(1) manual entry for information about the
\fBLESSCHARDEF\fR environment variable.)

.Uh "Storing the Contents"
The `\-store' switch tells \fImhn\fR to store the contents of the
named messages in \*(lqnative\*(rq (decoded) format.  Two things must
be determined: the directory to store the content, and the filenames.
Files are written in the directory given by the \fBnmh-storage\fR
profile entry,
.ne 6
e.g.,
.sp
.in +.5i
nmh-storage: /tmp
.in -.5i
.sp
If this entry isn't present,
the current working directory is used.

If the `\-auto' switch is given, then \fImhn\fR will check if the
message contains information indicating the filename that should be
used to store the content.  This information should be specified as the
attribute \*(lqname=filename\*(rq in the Content-Type header for the
content you are storing.  For security reasons, this filename will be
ignored if it begins with the character '/', '.', '|', or '!', or if it
contains the character '%'.  For the sake of security, this switch is
not the default, and it is recommended that you do NOT put the `\-auto'
switch in your \&.mh\(ruprofile file.

If the `\-auto' switch is not given (or is being ignored for
security reasons) then \fImhn\fR will look in the user's profile for
a \*(lqformatting string\*(rq to determine how the different contents
should be stored.  First, \fImhn\fR will look for an entry of the form:
.sp
.in +.5i
mhn-store-<type>/<subtype>
.in -.5i
.sp
to determine the formatting string.  If this isn't found, \fImhn\fR will
look for an entry of the form:
.sp
.in +.5i
mhn-store-<type>
.in -.5i
.sp
to determine the formatting string.

If the formatting string starts with a \*(lq+\*(rq character, then
content is stored in the named folder.  A formatting string consisting
solely of a \*(lq+\*(rq character is interpreted to be the current folder.

If the formatting string consists solely of a \*(lq-\*(rq character,
then the content is sent to the standard output.

If the formatting string starts with a '|', then the display string will
represent a command for \fImhn\fR to execute which should ultimately
store the content.  The content will be passed to the standard input of
the command.  Before the command is executed, \fImhn\fR will change to
the appropriate directory, and any escapes (given below) in the display
string will be expanded.

Otherwise the formatting string will represent a pathname in which to
store the content.  If the formatting string starts with a '/', then the
content will be stored in the full path given, else the file name will
be relative to the value of \fBnmh-storage\fR or the current working
directory.  Any escapes (given below) will be expanded, except for the
a-escape.

A command or pathname formatting string may contain the following escapes.
If the content isn't part of a multipart (of any subtype listed above)
content, the p-escapes are ignored.
.sp
.nf
.in +.5i
.ta \w'%P  'u
%a	Parameters from Content-type  (only valid with command)
%m	Insert message number
%P	Insert part number with leading dot
%p	Insert part number without leading dot
%t	Insert content type
%s	Insert content subtype
%%	Insert character %
.re
.in -.5i
.fi
.sp
If no formatting string is found, \fImhn\fR will check to see if the
content is application/octet-stream with parameter \*(lqtype=tar\*(rq.
If so, \fImhn\fR will choose an appropriate filename.  If the content

doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/mhn.c

/*
 * mhn.c -- display, list, cache, or store the contents of MIME messages
 *
 * $Id: mhn.c,v 1.1.1.1 1999-02-07 18:14:14 danw Exp $
 */

#include <h/mh.h>
#include <fcntl.h>
#include <h/signals.h>
#include <h/md5.h>
#include <errno.h>
#include <signal.h>
#include <zotnet/mts/mts.h>
#include <zotnet/tws/tws.h>
#include <h/mime.h>
#include <h/mhparse.h>
#include <h/mhcachesbr.h>

#ifdef HAVE_SYS_WAIT_H
# include <sys/wait.h>
#endif

/*
 * We allocate space for message names (msgs array)
 * this number of elements at a time.
 */
#define MAXMSGS  256


static struct swit switches[] = {
#define	AUTOSW                  0
    { "auto", 0 },
#define	NAUTOSW                 1
    { "noauto", 0 },
#define	CACHESW                 2
    { "cache", 0 },
#define	NCACHESW                3
    { "nocache", 0 },
#define	CHECKSW                 4
    { "check", 0 },
#define	NCHECKSW                5
    { "nocheck", 0 },
#define	HEADSW                  6
    { "headers", 0 },
#define	NHEADSW                 7
    { "noheaders", 0 },
#define	LISTSW                  8
    { "list", 0 },
#define	NLISTSW                 9
    { "nolist", 0 },
#define	PAUSESW                10
    { "pause", 0 },
#define	NPAUSESW               11
    { "nopause", 0 },
#define	SIZESW                 12
    { "realsize", 0 },
#define	NSIZESW                13
    { "norealsize", 0 },
#define	SERIALSW               14
    { "serialonly", 0 },
#define	NSERIALSW              15
    { "noserialonly", 0 },
#define	SHOWSW                 16
    { "show", 0 },
#define	NSHOWSW                17
    { "noshow", 0 },
#define	STORESW                18
    { "store", 0 },
#define	NSTORESW               19
    { "nostore", 0 },
#define	VERBSW                 20
    { "verbose", 0 },
#define	NVERBSW                21
    { "noverbose", 0 },
#define	FILESW                 22	/* interface from show */
    { "file file", 0 },
#define	FORMSW                 23
    { "form formfile", 0 },
#define	PARTSW                 24
    { "part number", 0 },
#define	TYPESW                 25
    { "type content", 0 },
#define	RCACHESW               26
    { "rcache policy", 0 },
#define	WCACHESW               27
    { "wcache policy", 0 },
#define VERSIONSW              28
    { "version", 0 },
#define	HELPSW                 29
    { "help", 4 },

/*
 * switches for debugging
 */
#define	DEBUGSW                30
    { "debug", -5 },

/*
 * switches for moreproc/mhlproc
 */
#define	PROGSW                 31
    { "moreproc program", -4 },
#define	NPROGSW                32
    { "nomoreproc", -3 },
#define	LENSW                  33
    { "length lines", -4 },
#define	WIDTHSW                34
    { "width columns", -4 },

/*
 * switches for mhbuild
 */
#define BUILDSW                35
    { "build", -5 },
#define NBUILDSW               36
    { "nobuild", -7 },
#define	EBCDICSW               37
    { "ebcdicsafe", -10 },
#define	NEBCDICSW              38
    { "noebcdicsafe", -12 },
#define	RFC934SW               39
    { "rfc934mode", -10 },
#define	NRFC934SW              40
    { "norfc934mode", -12 },
    { NULL, 0 }
};


extern int errno;

/* mhparse.c */
extern int checksw;
extern char *tmp;	/* directory to place temp files */

/* mhcachesbr.c */
extern int rcachesw;
extern int wcachesw;
extern char *cache_public;
extern char *cache_private;

/* mhshowsbr.c */
extern int pausesw;
extern int serialsw;
extern char *progsw;
extern int nolist;
extern int nomore;	/* flags for moreproc/header display */
extern char *formsw;

/* mhstoresbr.c */
extern int autosw;
extern char *cwd;	/* cache current working directory */

/* mhmisc.c */
extern int npart;
extern int ntype;
extern char *parts[NPARTS + 1];
extern char *types[NTYPES + 1];
extern int userrs;

int debugsw = 0;
int verbosw = 0;

/* The list of top-level contents to display */
CT *cts = NULL;

/*
 * variables for mhbuild (mhn -build)
 */
static int buildsw  = 0;
static int ebcdicsw = 0;
static int rfc934sw = 0;

/*
 * what action to take?
 */
static int cachesw = 0;
static int listsw  = 0;
static int showsw  = 0;
static int storesw = 0;

#define	quitser	pipeser

/* mhparse.c */
CT parse_mime (char *);

/* mhmisc.c */
int part_ok (CT, int);
int type_ok (CT, int);
void set_endian (void);
void flush_errors (void);

/* mhshowsbr.c */
void show_all_messages (CT *);

/* mhlistsbr.c */
void list_all_messages (CT *, int, int, int, int);

/* mhstoresbr.c */
void store_all_messages (CT *);

/* mhcachesbr.c */
void cache_all_messages (CT *);

/* mhfree.c */
void free_content (CT);

/*
 * static prototypes
 */
static RETSIGTYPE pipeser (int);


int
main (int argc, char **argv)
{
    int sizesw = 1, headsw = 1;
    int nummsgs, maxmsgs, msgnum, *icachesw;
    char *cp, *file = NULL, *folder = NULL;
    char *maildir, buf[100], **argp;
    char **arguments, **msgs;
    struct msgs *mp = NULL;
    CT ct, *ctp;
    FILE *fp;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    /*
     * Allocate the initial space to record message
     * names, ranges, and sequences.
     */
    nummsgs = 0;
    maxmsgs = MAXMSGS;
    if (!(msgs = (char **) malloc ((size_t) (maxmsgs * sizeof(*msgs)))))
	adios (NULL, "unable to allocate storage");

    /*
     * Parse arguments
     */
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
	    case AMBIGSW: 
		ambigsw (cp, switches);
		done (1);
	    case UNKWNSW: 
		adios (NULL, "-%s unknown", cp);

	    case HELPSW: 
		snprintf (buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			invo_name);
		print_help (buf, switches, 1);
		done (1);
	    case VERSIONSW:
		print_version(invo_name);
		done (1);

	    case AUTOSW:
		autosw++;
		continue;
	    case NAUTOSW:
		autosw = 0;
		continue;

	    case CACHESW:
		cachesw++;
		continue;
	    case NCACHESW:
		cachesw = 0;
		continue;

	    case RCACHESW:
		icachesw = &rcachesw;
		goto do_cache;
	    case WCACHESW:
		icachesw = &wcachesw;
do_cache:
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		switch (*icachesw = smatch (cp, caches)) {
		case AMBIGSW:
		    ambigsw (cp, caches);
		    done (1);
		case UNKWNSW:
		    adios (NULL, "%s unknown", cp);
		default:
		    break;
		}
		continue;

	    case CHECKSW:
		checksw++;
		continue;
	    case NCHECKSW:
		checksw = 0;
		continue;

	    case HEADSW:
		headsw = 1;
		continue;
	    case NHEADSW:
		headsw = 0;
		continue;

	    case LISTSW:
		listsw = 1;
		continue;
	    case NLISTSW:
		listsw = 0;
		continue;

	    case PAUSESW:
		pausesw = 1;
		continue;
	    case NPAUSESW:
		pausesw = 0;
		continue;

	    case SERIALSW:
		serialsw = 1;
		continue;
	    case NSERIALSW:
		serialsw = 0;
		continue;

	    case SHOWSW:
		showsw = 1;
		continue;
	    case NSHOWSW:
		showsw = 0;
		continue;

	    case SIZESW:
		sizesw = 1;
		continue;
	    case NSIZESW:
		sizesw = 0;
		continue;

	    case STORESW:
		storesw = 1;
		continue;
	    case NSTORESW:
		storesw = 0;
		continue;

	    case PARTSW:
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		if (npart >= NPARTS)
		    adios (NULL, "too many parts (starting with %s), %d max",
			   cp, NPARTS);
		parts[npart++] = cp;
		continue;

	    case TYPESW:
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		if (ntype >= NTYPES)
		    adios (NULL, "too many types (starting with %s), %d max",
			   cp, NTYPES);
		types[ntype++] = cp;
		continue;

	    case FILESW:
		if (!(cp = *argp++) || (*cp == '-' && cp[1]))
		    adios (NULL, "missing argument to %s", argp[-2]);
		file = *cp == '-' ? cp : path (cp, TFILE);
		continue;

	    case FORMSW:
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		if (formsw)
		    free (formsw);
		formsw = getcpy (etcpath (cp));
		continue;

	    /*
	     * Switches for moreproc/mhlproc
	     */
	    case PROGSW:
		if (!(progsw = *argp++) || *progsw == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		continue;
	    case NPROGSW:
		nomore++;
		continue;

	    case LENSW:
	    case WIDTHSW:
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		continue;

	    /*
	     * Switches for mhbuild
	     */
	    case BUILDSW:
		buildsw = 1;
		continue;
	    case NBUILDSW:
		buildsw = 0;
		continue;
	    case RFC934SW:
		rfc934sw = 1;
		continue;
	    case NRFC934SW:
		rfc934sw = -1;
		continue;
	    case EBCDICSW:
		ebcdicsw = 1;
		continue;
	    case NEBCDICSW:
		ebcdicsw = -1;
		continue;

	    case VERBSW: 
		verbosw = 1;
		continue;
	    case NVERBSW: 
		verbosw = 0;
		continue;
	    case DEBUGSW:
		debugsw = 1;
		continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = path (cp + 1, *cp == '+' ? TFOLDER : TSUBCWF);
	} else {
	    /*
	     * Check if we need to allocate more space
	     * for message names/ranges/sequences.
	     */
	    if (nummsgs >= maxmsgs) {
		maxmsgs += MAXMSGS;
		if (!(msgs = (char **) realloc (msgs,
			(size_t) (maxmsgs * sizeof(*msgs)))))
		    adios (NULL, "unable to reallocate msgs storage");
	    }
	    msgs[nummsgs++] = cp;
	}
    }

    /* null terminate the list of acceptable parts/types */
    parts[npart] = NULL;
    types[ntype] = NULL;

    set_endian ();

    if ((cp = getenv ("MM_NOASK")) && !strcmp (cp, "1")) {
	nolist  = 1;
	listsw  = 0;
	pausesw = 0;
    }

    /*
     * Check if we've specified an additional profile
     */
    if ((cp = getenv ("MHN"))) {
	if ((fp = fopen (cp, "r"))) {
	    readconfig ((struct node **) 0, fp, cp, 0);
	    fclose (fp);
	} else {
	    admonish ("", "unable to read $MHN profile (%s)", cp);
	}
    }

    /*
     * Read the standard profile setup
     */
    if ((fp = fopen (cp = etcpath ("mhn.defaults"), "r"))) {
	readconfig ((struct node **) 0, fp, cp, 0);
	fclose (fp);
    }

    /* Check for public cache location */
    if ((cache_public = context_find (nmhcache)) && *cache_public != '/')
	cache_public = NULL;

    /* Check for private cache location */
    if (!(cache_private = context_find (nmhprivcache)))
	cache_private = ".cache";
    cache_private = getcpy (m_maildir (cache_private));

    /*
     * Cache the current directory before we do any chdirs()'s.
     */
    cwd = getcpy (pwd());

    /*
     * Check for storage directory.  If specified,
     * then store temporary files there.  Else we
     * store them in standard nmh directory.
     */
    if ((cp = context_find (nmhstorage)) && *cp)
	tmp = concat (cp, "/", invo_name, NULL);
    else
	tmp = add (m_maildir (invo_name), NULL);

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    /*
     * Process a mhn composition file (mhn -build)
     */
    if (buildsw) {
	char *vec[MAXARGS];
	int vecp;

	if (showsw || storesw || cachesw)
	    adios (NULL, "cannot use -build with -show, -store, -cache");
	if (nummsgs < 1)
	    adios (NULL, "need to specify a %s composition file", invo_name);
	if (nummsgs > 1)
	    adios (NULL, "only one %s composition file at a time", invo_name);

	vecp = 0;
	vec[vecp++] = "mhbuild";

	if (ebcdicsw == 1)
	    vec[vecp++] = "-ebcdicsafe";
	else if (ebcdicsw == -1)
	    vec[vecp++] = "-noebcdicsafe";

	if (rfc934sw == 1)
	    vec[vecp++] = "-rfc934mode";
	else if (rfc934sw == -1)
	    vec[vecp++] = "-norfc934mode";

	vec[vecp++] = msgs[0];
	vec[vecp] = NULL;

	execvp ("mhbuild", vec);
	fprintf (stderr, "unable to exec ");
	_exit (-1);
    }

    /*
     * Process a mhn composition file (old MH style)
     */
    if (nummsgs == 1 && !folder && !npart && !cachesw
	&& !showsw && !storesw && !ntype && !file
	&& (cp = getenv ("mhdraft"))
	&& strcmp (cp, msgs[0]) == 0) {

	char *vec[MAXARGS];
	int vecp;

	vecp = 0;
	vec[vecp++] = "mhbuild";

	if (ebcdicsw == 1)
	    vec[vecp++] = "-ebcdicsafe";
	else if (ebcdicsw == -1)
	    vec[vecp++] = "-noebcdicsafe";

	if (rfc934sw == 1)
	    vec[vecp++] = "-rfc934mode";
	else if (rfc934sw == -1)
	    vec[vecp++] = "-norfc934mode";

	vec[vecp++] = cp;
	vec[vecp] = NULL;

	execvp ("mhbuild", vec);
	fprintf (stderr, "unable to exec ");
	_exit (-1);
    }

    if (file && nummsgs)
	adios (NULL, "cannot specify msg and file at same time!");

    /*
     * check if message is coming from file
     */
    if (file) {
	if (!(cts = (CT *) calloc ((size_t) 2, sizeof(*cts))))
	    adios (NULL, "out of memory");
	ctp = cts;

	if ((ct = parse_mime (file)));
	    *ctp++ = ct;
    } else {
	/*
	 * message(s) are coming from a folder
	 */
	if (!nummsgs)
	    msgs[nummsgs++] = "cur";
	if (!folder)
	    folder = getfolder (1);
	maildir = m_maildir (folder);

	if (chdir (maildir) == NOTOK)
	    adios (maildir, "unable to change directory to");

	/* read folder and create message structure */
	if (!(mp = folder_read (folder)))
	    adios (NULL, "unable to read folder %s", folder);

	/* check for empty folder */
	if (mp->nummsg == 0)
	    adios (NULL, "no messages in %s", folder);

	/* parse all the message ranges/sequences and set SELECTED */
	for (msgnum = 0; msgnum < nummsgs; msgnum++)
	    if (!m_convert (mp, msgs[msgnum]))
		done (1);
	seq_setprev (mp);	/* set the previous-sequence */

	if (!(cts = (CT *) calloc ((size_t) (mp->numsel + 1), sizeof(*cts))))
	    adios (NULL, "out of memory");
	ctp = cts;

	for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++) {
	    if (is_selected(mp, msgnum)) {
		char *msgnam;

		msgnam = m_name (msgnum);
		if ((ct = parse_mime (msgnam)))
		    *ctp++ = ct;
	    }
	}
    }

    if (!*cts)
	done (1);

    /*
     * You can't give more than one of these flags
     * at a time.
     */
    if (showsw + listsw + storesw + cachesw > 1)
	adios (NULL, "can only use one of -show, -list, -store, -cache at same time");

    /* If no action is specified, assume -show */
    if (!listsw && !showsw && !storesw && !cachesw)
	showsw = 1;

    userrs = 1;
    SIGNAL (SIGQUIT, quitser);
    SIGNAL (SIGPIPE, pipeser);

    /*
     * Get the associated umask for the relevant contents.
     */
    for (ctp = cts; *ctp; ctp++) {
	struct stat st;

	ct = *ctp;
	if (type_ok (ct, 1) && !ct->c_umask) {
	    if (stat (ct->c_file, &st) != NOTOK)
		ct->c_umask = ~(st.st_mode & 0777);
	    else
		ct->c_umask = ~m_gmprot();
	}
    }

    /*
     * List the message content
     */
    if (listsw)
	list_all_messages (cts, headsw, sizesw, verbosw, debugsw);

    /*
     * Store the message content
     */
    if (storesw)
	store_all_messages (cts);

    /*
     * Cache the message content
     */
    if (cachesw)
	cache_all_messages (cts);

    /*
     * Show the message content
     */
    if (showsw)
	show_all_messages (cts);

    /* Now free all the structures for the content */
    for (ctp = cts; *ctp; ctp++)
	free_content (*ctp);

    free ((char *) cts);
    cts = NULL;

    /* If reading from a folder, do some updating */
    if (mp) {
	context_replace (pfolder, folder);/* update current folder  */
	seq_setcur (mp, mp->hghsel);	  /* update current message */
	seq_save (mp);			  /* synchronize sequences  */
	context_save ();		  /* save the context file  */
    }

    done (0);
    /* NOTREACHED */
}


static RETSIGTYPE
pipeser (int i)
{
    if (i == SIGQUIT) {
	unlink ("core");
	fflush (stdout);
	fprintf (stderr, "\n");
	fflush (stderr);
    }

    done (1);
    /* NOTREACHED */
}


void
done (int status)
{
    CT *ctp;

    if ((ctp = cts))
	for (; *ctp; ctp++)
	    free_content (*ctp);

    exit (status);
}
